src/shared/extension-release.c

```diff
@@ -48,6 +49,15 @@ int extension_release_validate(
                 }
         }
 
+        /* When the arch field is present and not 'any' it must match the host - for now just look at uname but in
+         * the future we could check if the kernel binfmt has something like qemu-static set up for the arch */
+        extension_arch = strv_env_pairs_get(extension_release, "ARCH");
```

Comment:
hmm, this is a new field. Do we need that? What kind of images are you operating with here? Note that if you use GPT-style disk images with verity and stuff you have that same info already in the GPT partition type uuid....

I mean, I can see how this would be useful in directory/tarball images.

If we add this, I think this should be added to the full os-release primarily, and then be inherited into the extension-release thing from there.

Reply:
Also, not really a fan of unnecessary abbreviations for things. There's nothing wrong with just calling the field $ARCHITECTURE

Reply:
My use case are squashfs images because they are easy to build (no loop device setup needed, does even work from containers) but they don't have GPT partitions. Anyway, I don't think that the GPT is the best way to encode this and it's better to store the requirements in the release file at one place.
The os-release file of the host doesn't really need to contain this field from my perspective because what I am interested in is whether the kernel can execute these binaries, not whether the distribution uses has the libraries/binaries (e.g., a multiarch debian would have both x86 and x86-64). This reminds me that besides binfmt we can also handle the case of 32 bit x86 binaries being executable on a 64 bit kernel but that's not the most important thing right now.
Using `ARCHITECTURE` is good for me, too.

Reply:
> My use case are squashfs images because they are easy to build (no loop device setup needed, does even work from containers) but they don't have GPT partitions.

Hmm, what? You are mounting squashfs images without loopback devices? How does that work?  Through some fuse shit?

Reply:
> The os-release file of the host doesn't really need to contain this field from my perspective because what I am interested in is whether the kernel can execute these binaries, not whether the distribution uses has the libraries/binaries (e.g., a multiarch debian would have both x86 and x86-64). 

Well, it should be an optional field for environments where this is desirable. I am pretty sure extension-release and os-release should really be developed side-by-side, and hence have the same fields, and extension-release should "import" fields from os-release wherever that makes sense.

os-release is used for nspawn-style container images and for portable services too. It might make sense to encode the architecture there, if not operating on a GPT image.

I understand you are using raw squashfs without a gpt envelope. Do note that once you intend to support verity (i.e. signing your stuff sensibly) you'll notice that this will be too simple.

> This reminds me that besides binfmt we can also handle the case of 32 bit x86 binaries being executable on a 64 bit kernel but that's not the most important thing right now.

Sure, but that's a question orthogonal to the question if ARCHITECTURE= makes sense to declare what an image is for. I think it does make sense for images not doing GPT. And hence if this is added at all it should be added and defined for "os-release", and then imported into the spec for "extension-release". But it should be an optional field either way, since on GPT images it's kinda useless, and there are multi-arch images thinkable.

Reply:
> > My use case are squashfs images because they are easy to build (no loop device setup needed, does even work from containers) but they don't have GPT partitions.
> 
> Hmm, what? You are mounting squashfs images without loopback devices? How does that work? Through some fuse mess?

To `build` images - we are doing the same. Building a squashfs is an entirely-userspace operation requiring no kernel involvement, root privileges, access to blockdev/loopdev, and so on, which is very very convenient.

Reply:
> I understand you are using raw squashfs without a gpt envelope. Do note that once you intend to support verity (i.e. signing your stuff sensibly) you'll notice that this will be too simple.

We actually do verity with squashfs images just fine. Yes it's annoying to carry over 4 artifacts instead of one, but it is not a show-stopper.

Reply:
> To `build` images - we are doing the same. Building a squashfs is an entirely-userspace operation requiring no kernel involvement, root privileges, access to blockdev/loopdev, and so on, which is very very convenient.

But so does wrapping them in GPT? What am I missing?

Reply:
> We actually do verity with squashfs images just fine. Yes it's annoying to carry over 4 artifacts instead of one, but it is not a show-stopper.

For new stuff I'd advise maybe not making your life unnecessarily complex.

Reply:
> > To `build` images - we are doing the same. Building a squashfs is an entirely-userspace operation requiring no kernel involvement, root privileges, access to blockdev/loopdev, and so on, which is very very convenient.
> 
> But so does wrapping them in GPT? What am I missing?

Without a loop device? How do you copy the squashfs to the partition?

> > We actually do verity with squashfs images just fine. Yes it's annoying to carry over 4 artifacts instead of one, but it is not a show-stopper.
> 
> For new stuff I'd advise maybe not making your life unnecessarily complex.

Sure, depending on requirements - but the point is that verity does fully work in that case too

Reply:
> Without a loop device? How do you copy the squashfs to the partition?

You can either use a ready-made tool like systemd-repart which will do that for you. You can alternatively script something simple with sfdisk+dd if you like, it's not that hard: just define the minimal gpt table in sfdisk textual representation, with the right size, then call sfdisk to make an image file from it, then use sfdisk again, to dump the resulting full gpt table in text format, parse out the start offset, and dd the squashfs into that offset. dd/sfdisk operate just fine on regular files. Works without any privs.

Reply:
Right - or you do `mksquashfs dir/` and then use the rest of the time to have a nap instead :-)

Reply:
> Hmm, what? You are mounting squashfs images without loopback devices? How does that work? Through some fuse mess?

I meant creation of squashfs images with mksquashfs makes extensions easy to build because no mounting is involved.

Reply:
Does making `ARCHITECTURE` a field for `os-release` also mean that we should use it instead of `uname`'s arch to match against the extension's field? In this case it wouldn't be very useful for distro-independent images because each distro's `os-release` would need to have it then. I'm fine adding it to the spec if we don't use the host's `os-release` `ARCHITECTURE` in the sysext check but use the `uname` arch instead. Where does the spec for `os-release` live?

Reply:
No, I think ARCHITECTURE is decoration, not more. i.e. a property you can use if you have nothing better, but things like uname() should always win, because they *actually* say what is in effect.

Reply:
Added docs for the field to the os-release man page

Reply:
> Right - or you do `mksquashfs dir/` and then use the rest of the time to have a nap instead :-)

Well, it's called technical debt, if you lock yourself into a world, where any extension means shipping/downloading more stuff. I mean, you are patching systemd here to support something you might as well achieve by just wrapping your stuff in a proper GPT envelope, that makes things auto-detectable, recognizable and self-descriptive. The fact that this PR was filed to basically add inside the image what we already support via the GPT envelope shows that you are making your changes at the wrong place.

I mean, adding more C code to get a crappier result where you could just add 4 lines of shell instead to give you a better result sounds like an awful choice to me.

@pothos is adding new infra to flatcaros, and I'd recommend to do that in way that ensures you keep things simple and extensible for the future, so that you have images that carry verity and pkcs7 inside them and whatever else needs to be added sooner or later. Doing this in sidecar files is just a mess, you create an artificial need to download many resources, sync them up and so on where you could just have a single blob instead, and that in an extensible way.

Reply:
Maybe to add, I do this here with end users or third-party software distributors in mind. My main motivation for having ARCHITECTURE in the release file was to have everything in one place and also cover the squashfs/dir case. I agree, dm-verity is great, GPT partition GUIDs are also great (one could even bundle multiple architectures into a single image, each having its own partition - but it won't work right now as written below), and for the reasons you mentioned, both should be recommended over plain fs images. For beginners a quick squashfs is nice and I imagine people may prefer if over the verity dance - if we want to push users to verity, we have to improve the docs a lot from my point of view, even how to use GPT with sysext is not really documented (i.e., use the root fs GUID of the arch for the partition type and not something else).
Related to this: Currently the GPT GUID for the partition type is not checked against the host architecture for a loading the sysext image, so as of now it's not an alternative to the ARCHITECTURE field, should I open an issue about that?

Reply:
Filed an issue for that here: https://github.com/systemd/systemd/issues/24318

Reply:
> Related to this: Currently the GPT GUID for the partition type is not checked against the host architecture for a loading the sysext image, so as of now it's not an alternative to the ARCHITECTURE field, should I open an issue about that?

Well, it's used for searching the root fs in the image. We wouldn't even find the partition in the image if the gpt type uuid wouldn't exist.